### PR TITLE
hyperHTML fully refactored - toward V1

### PR DESCRIPTION
**Pull request by WebReflection**
- compartible by default with Firefox < 55
  - improved IE compatibility
  - 2X better performance than previous version
  - avoided any redundant operation
  - new algorithm to map node once per template literal
  - better memory management
  - extremely improved `hyperHTML.adopt(el)` reliability
  - adopted nodes share same pre-parsed literal of bound one
  - improved virtual adoption test coverage

TODO:
  - hyperHTML.wire(any?, 'adopt')

### CODE REVIEW

**Comment by WebReflection** [comment]
Hey @marcoscaceres, if you have time please have a look.

I wouldn't even compare current code with previous one, because it has been fully rewritten.

I also have plans for the bound wires but I'd like to align current status sooner than later with the master.

Bound wires will need some extra refactoring of `setContent` methods but I have already a way better surrounding structure than I did with previous code.

My lib size estimation by the end of fully supported adopted wires is around 3.5KB minzipped.

I know it's quite a lot compare to nearly 2KB but I think it's worth exploring the adoption pattern more because my previous tests showed extremely good result for both performance and progressive enhancement (adopting on demand while typing an input without losing focus ... now that's something)

#### Other noticeable improvements:

  * I don't want to declare once on the top each var and I don't care hoisting. The VS Code basic linter here helped a lot and I think it's overall a cleaner code so I hope you'll appreciate this
  * the amount of duplicated code, beside `for` loops definition, has been reduced to the minimum
  * whenever available, the platform is always preferred over ugly hacks (no EXPANDO anymore, just Map and WeakMap)
  * Firefox performance has been kept under control and are now "_as good as these could be_"
  * Firefox also now is supported by default and ugly-patched via UA sniff because I cannot feature detect new syntax such template literals ... very ugly but it works
  * the test page and other examples have been tested against Firefox, Safari, Chrome, and Edge + Mobile couter  part, without the need to transpile anything

Maybe more ... right now I'm just quite excited about this!

Hope you'll like little changes here and there.

**Comment by marcoscaceres** [comment]
Will take a look first thing tomorrow (about 12 hours)

**Comment by albertosantini** [comment]
I tried this hyperHTML version (it is the release contained in this PR):

https://rawgit.com/WebReflection/hyperHTML/b6af7885fa1bd8c1afab6214c3b44af2c1e20410/hyperhtml.js

In the branch `hyperHTML` of Argo, the porting from AngularJS (almost completed):

https://github.com/albertosantini/argo/tree/hyperHTML 

It seems it doesn't break anything and the app works fine.

**Comment by WebReflection** [comment]
Right before @marcoscaceres wakes up for the review :smile: 

The algo shown in `abstract.md` file has the entry point reversed.

Logically speaking you would expect that the most important thing to know is if the `templateObject` / `statics` parameter is known, and if it's not it should be known once and once only.

Well, turns out, pragmatically speaking, the first operation I need to do is to understand if the current function bound context is known instead, because in such case there's nothing else to, while if it's not known there is to check if the `templateObject` is known first, and then eventually apply it to the current context.

It's a little detail but explains why the first operation in `hyperHTML` or `.wire` is to check for the context `WeakMap` instead of `templateObject` first.

**Comment by marcoscaceres** [comment]
@WebReflection this is exciting to see! Because of the significant number of changes I want to go through it in detail. I usually review about ~100 lines per hour, so I'll try to do this in chunks and send feedback as I go over the next 2-3 days. Hope that's ok time-wise.

File: hyperhtml.js [hunk 1]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+    return function text(newValue) {
+      if (oldValue !== newValue) {
+        oldValue = newValue;
+        node.textContent = newValue;
+      }
+    };
+  }
+
+  // `<li>a</li>${'virtual'}<li>c</li>`
+  function setVirtualContent(node, childNodes) {
```
Comment (marcoscaceres) [inline hyperhtml.js]: Maybe add a bit more of clarification of what you mean by "virtual", as it's an overloaded word.

File: hyperhtml.js [hunk 2]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
-                value = value.slice(i);
-                appendNodes(fragment, value);
-                parentNode.insertBefore(fragment, node);
-                childNodes = childNodes.slice(0, i).concat(value);
+              var type = typeof value[0];
+              if (type === 'string') {
+                any(value.join(''));
+              } else {
+                var i = indexOfDifferences(childNodes, value);
+                if (-1 < i) {
```
Comment (marcoscaceres) [inline hyperhtml.js]: Nit: Same comment as above about idiomaticness. 

```JS
var i = indexOfDifferences(childNodes, value) 
if(i !== -1) {}
```

File: hyperhtml.js [hunk 3]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
-    var oldValue;
-    return function text(newValue) {
-      if (oldValue !== newValue) {
-        node.textContent = (oldValue = newValue);
+  // ---------------------------------------------
+  // DOM Traversing
+  // ---------------------------------------------
+
+  // look for attributes that contains the comment text
+  function attributesSeeker(node, paths) {
```
Comment (marcoscaceres) [inline hyperhtml.js]: NOTE TO @marcoscacers. Stopped here on day 1.

**Comment by marcoscaceres** [comment]
@WebReflection, maybe wait until I finish review before pushing new changes? Just so I don't need to backtrack.

File: hyperhtml.js [hunk 4]
```diff
@@ -519,193 +853,34 @@ var hyperHTML = (function () {'use strict';
...
   // ⚡️ ️️The End ➰
-  // -------------------------
+  // ---------------------------------------------
   return hyperHTML;
 
-}());
+}(document));
 
 // umd.KISS
 try { module.exports = hyperHTML; } catch(o_O) {}
```
Comment (WebReflection) [inline hyperhtml.js]: it's death, as it is bower

File: hyperhtml.js [hunk 5]
```diff
@@ -1,56 +1,43 @@
-var hyperHTML = (function () {'use strict';
+var hyperHTML = (function (document) {'use strict';
```
Comment (WebReflection) [inline hyperhtml.js]: the only reason it's there is for minifiers but it makes sense to give it a better name.

File: hyperhtml.js [hunk 6]
```diff
@@ -1,56 +1,43 @@
...
-  //
-  // var render = hyperHTML.adopt(document.body);
-  // render`<div adopted="${true}"> ${'new text'} </div>`;
+  // ---------------------------------------------
+  // hyperHTML Public API
+  // ---------------------------------------------
+
+  // The document must be swap-able at runtime.
+  // Needed by both basicHTML and nativeHTML
+  hyperHTML.document = document;
```
Comment (WebReflection) [inline hyperhtml.js]: it's swapped in nativeHTML. The swapping is needed elsewhere, not in this project.
https://github.com/WebReflection/nativeHTML/blob/master/nativehtml.js#L28-L34

Comment (WebReflection) [inline hyperhtml.js]: also basicHTML might need swapping, and this project is compatible with it:
https://github.com/WebReflection/basicHTML/blob/master/test/hyperhtml.js

File: hyperhtml.js [hunk 7]
```diff
@@ -1,56 +1,43 @@
...
+
+  // hyperHTML.bind(el) ⚡️
+  function hyperHTML(template) {
+    var hyper = hypers.get(this);
+    if (
+      !hyper ||
+      hyper.template !== (FF ? unique(template) : template)
+    ) {
+      var fragment = upgrade.apply(this, arguments);
+      if (notAdopting) {
```
Comment (WebReflection) [inline hyperhtml.js]: I rather not touch arguments at all and `notAdopting` is private in the scope (just clarifying since you said it's global)

You can see `notAdopting` as the only meaningful _state_ in `hyperHTML`

File: hyperhtml.js [hunk 8]
```diff
@@ -60,129 +47,105 @@ var hyperHTML = (function () {'use strict';
...
+        var childNodes = parent.childNodes;
+        if (childNodes[0] === child) {
+          removeNodeList(childNodes, 1);
           break;
-      }
+        }
+        resetAndPopulate(parent, child);
+        break;
+      case DOCUMENT_FRAGMENT_NODE:
+        if (-1 < indexOfDifferences(parent.childNodes, child.childNodes)) {
```
Comment (WebReflection) [inline hyperhtml.js]: would `fn(...) > -1` work too?

File: hyperhtml.js [hunk 9]
```diff
@@ -60,129 +47,105 @@ var hyperHTML = (function () {'use strict';
...
-  //
-  // var render = hyperHTML.bind(node);
-  // function update() {
-  //  render`template`;
-  // }
+  // `<div>${'any'}</div>`
   function setAnyContent(node) {
     return function any(value) {
-      switch (typeof value) {
+      var type = typeof value;
```
Comment (WebReflection) [inline hyperhtml.js]: you are comparing with previous function but it's been fully rewritten.
```js
  // `<div>${'any'}</div>`
  function setAnyContent(node) {
    return function any(value) {
      var type = typeof value;
      switch (type) {
        case 'string':
        case 'number':
        case 'boolean':
          node.innerHTML = value;
          break;
        default:
          if (isArray(value)) {
            var length = value.length;
            if (length === 1) {
              any(value[0]);
            } else {
              if (0 < length) type = typeof value[0];
              if (type === 'string') {
                any(value.join(''));
              } else {
                var i = indexOfDifferences(node.childNodes, value);
                if (-1 < i) updateViaArray(node, value, i);
              }
            }
          } else {
            populateNode(node, value);
          }
          break;
      }
    };
  }
```
I don't see any issue there

File: hyperhtml.js [hunk 10]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+      node = attribute.ownerElement,
       isEvent = name.slice(0, 2) === 'on',
       isSpecial = name in node && !SHOULD_USE_ATTRIBUTE.test(name),
       oldValue
     ;
-    if (isSpecial) node.removeAttribute(name);
+    if (isSpecial) removeAttributes.push(node, name);
     return isSpecial ?
       function specialAttr(newValue) {
         if (oldValue !== newValue) {
```
Comment (WebReflection) [inline hyperhtml.js]: makes sense

File: hyperhtml.js [hunk 11]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
-              newValue;
+          if (isEvent && newValue && ('handleEvent' in newValue)) {
+            handleEvent(node, 'add', name, newValue);
           } else {
             node[name] = newValue;
           }
         }
       } :
       function attr(newValue) {
         if (oldValue !== newValue) {
```
Comment (WebReflection) [inline hyperhtml.js]: actually, after changing it, I've realized I don't have any empty return anywhere else ... leaving as it is for now :stuck_out_tongue:

File: hyperhtml.js [hunk 12]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
-  //       update of them all.
-  function setVirtualContent(node) {
-    var
-      fragment = node.ownerDocument.createDocumentFragment(),
-      childNodes = []
-    ;
+  // `<div> ${'text'} </div>`
+  function setTextContent(node) {
+    var oldValue;
+    return function text(newValue) {
```
Comment (WebReflection) [inline hyperhtml.js]: uhm ... names of the update functions reflect the type. `attr` `any` and `text` ... this could also just be anonymous. Maybe it's just OK like it is?

Comment (WebReflection) [inline hyperhtml.js]: P.S. virtual is reserved word IIRC which is why `setVirtualContent` returns a `function any` but I am changing it now to return `anyVirtual`

File: hyperhtml.js (hunk 1 above)
Comment (WebReflection) [inline hyperhtml.js]: I've decided to avoid too much documentation in code and rather write proper documentation in the repo or in the website but I agree the code should be as much self-explanatory as possible.

Here the thing, when `hyperHTML` updates a node it's rather it's _text_, it's value (if _attribute_), or it's content as _any_ content. Then you have partial updates in between nodes, after a node, or before a node.

`<br>${'virtual'}`
`${'virtual'}<hr>`
`<br>${'virtual'}<hr>`

These are cases where the target node is not actually really existing, it's made up with arbitrary amount of nodes, text, html, and any sort of content but it's not a node when such content is empty (as empty string or empty array).

Hence the usage of _virtual_ content: it's not really there, but if populated, it's there.

The implementation uses a special comment node and an array of referenced elements to understand what should be removed and where new content, if any, should be re-appended.

As summary: do you have any better word to describe this situation?

File: hyperhtml.js (hunk 2 above)
Comment (WebReflection) [inline hyperhtml.js]: you like typing, don't you ? :smile:

File: hyperhtml.js [hunk 13]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+          if (SHOULD_USE_ATTRIBUTE.test(node.nodeName) && child.textContent === UIDC) {
+            paths.push(Path('text', node));
+          }
+          break;
+      }
+    }
+  }
 
-  // -------------------------
+  // ---------------------------------------------
```
Comment (WebReflection) [inline hyperhtml.js]: all sections are alphabetically ordered by topic like `Costants` but yeah, I'll move them up

Comment (WebReflection) [inline hyperhtml.js]: I've placed them right under the public API which looks like a more appropriate place. No costant is used in the public API anyway so there shouldn't be confusion

File: hyperhtml.js [hunk 14]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+  // without assuming Node is globally available
+  // since this project is used on the backend too
+  var ELEMENT_NODE = 1;
+  var ATTRIBUTE_NODE = 2;
+  var TEXT_NODE = 3;
+  var COMMENT_NODE = 8;
+  var DOCUMENT_FRAGMENT_NODE = 11;
+
+  var GET_ATTRIBUTE_NODE = 'getAttributeNode';
+  var SHOULD_USE_ATTRIBUTE = /^style$/i;
```
Comment (WebReflection) [inline hyperhtml.js]: nope, the list might be bigger than just style. `style` is one of those attributes that should not be used directly like `onclick` or `disabled` even if in the prototype.

I am sure there are other exceptions like style but I haven't fully investigated so until I'm sure 100% style is the only exception I'd like to keep the name open.

File: hyperhtml.js [hunk 15]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+  // the resulting template document fragment
+  function createFragment(node, html) {
+    var fragment;
+    var document = node.ownerDocument;
+    var container = document.createElement('template');
+    var hasContent = 'content' in container;
+    var fallback = false;
+    if (!hasContent) {
+      fragment = createDocumentFragment(document);
+      fallback = /^[^\S]*?<(t(?:head|body|foot|r|d|h))/i.test(html);
```
Comment (WebReflection) [inline hyperhtml.js]: it's not `table` needing special handling, it's `tr` or `td`.

You cannot `node.innerHTML = '<td></td>'` and retrieve that back. This is a DOM specification limitation with regular innerHTML and I'm not sure what to write about it.

File: hyperhtml.js [hunk 16]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
     }
-    virtual = parentNode;
-    while (parentNode = parentNode.parentNode) {
-      map.unshift('children', map.indexOf.call(parentNode.children, virtual));
-      virtual = parentNode;
+    if (fallback) {
```
Comment (WebReflection) [inline hyperhtml.js]: OK, I'll put a comment but this is a fallback in two cases: there is no HTMLTemplateElement in the engine, the content is a partial table internal content that cannot be created via innerHTML

tableFallback would be even more deceptive, because it's a "don't know about tempalte" problem too.

Comment (WebReflection) [inline hyperhtml.js]: scratch that, the content is not relevant. Putting a comment and changing name.

File: hyperhtml.js [hunk 17]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+    return fragment;
+  }
+
+  // given a document, it does what it says
+  function createDocumentFragment(document) {
+    return document.createDocumentFragment();
+  }
+
+  // given a node, it does what is says
+  function createText(node, text) {
```
Comment (WebReflection) [inline hyperhtml.js]: agreed for the fragment but this one plays nice in few places
```js
              if (before.length) {
                fragment.appendChild(createText(parentNode, before));
              }
              target = fragment.appendChild(createText(parentNode, ''));
              if (after.length) {
                fragment.appendChild(createText(parentNode, after));
              }
```

File: hyperhtml.js (hunk 9 above)
Comment (WebReflection) [inline hyperhtml.js]: actually I see what you meant

Comment (WebReflection) [inline hyperhtml.js]: well, not recycling the variable would branch out the current logic. I need to know the type of the first element of the array, if any, and if string simply invoke as such.

In the coming version with adopted wires that type can also be `function` instead of `string` and if I have to branch out the case with length and type plus the else which can be needed anyway I think I'll add complexity for no real gain.

Leaving like this for now. Will eventually improve once next changes are in.

File: hyperhtml.js [hunk 18]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
-    for (i = 0, length = map.length; i < length; i++) {
-      switch(name = map[i++]) {
-        case 'getAttributeNode':
-          parentNode = live[name](map[i]);
-          live = parentNode || {ownerElement: live, name: map[i]};
+    return fragment;
+  }
+
+  // given a document, it does what it says
+  function createDocumentFragment(document) {
```
Comment (WebReflection) [inline hyperhtml.js]: done

**Comment by WebReflection** [comment]
Hey @marcoscaceres I've just realized `hyperHTML` is broken when it comes to SVG nodes and since version `0.9.0`.

Luckily at the talk I've used a CodePen with such outdated version but I'd like to release ASAP this refactoring at this point since it (now) fixes that issue too.

Can you please give me an ETA to finish the review or tell me if it's OK if I release it (so far all points are aesthetics and styles but no bugs in the 100% code covered code) and do the writing / naming / styling review after ?

It's not that I need to rush too much but I am talking about this stuff and showing this stuff these days and all demos with SVGs are unusable :disappointed: 

Thanks for any outcome and sorry for the push

**Comment by marcoscaceres** [comment]
No problem if you need to release. I will continue tomorrow.

**Comment by WebReflection** [comment]
Actually there's more broken than I thought ... very weird

**Comment by albertosantini** [comment]
@WebReflection well, about svg it is working in my use case.

https://github.com/albertosantini/argo/blob/hyperHTML/src/client/app/components/sl-chart/sl-chart.template.js

**Comment by WebReflection** [comment]
@albertosantini I'm not sure it does and the problem is with attributes **and** with the new logic that creates a fragment and it clones it but it doesn't create a fragment with **real** SVG elements in it.

Anyway, as side note, why would you do this?
```js
    static update(render) {
        /* eslint indent: off */
        return render`${hyperHTML.wire(render, "svg")`
            <svg class="sl mw3"></svg>`
        }`;
    }
```
instead of just this?
```js
    static update(render) {
        /* eslint indent: off */
        return hyperHTML.wire(render, "svg")`
            <svg class="sl mw3"></svg>`;
    }
```

**Comment by WebReflection** [comment]
@marcoscaceres you know what? I've fixed everything locally but since it's been in a very bad shape (SVG speaking) for long time I'll just wait your review to publish this.

Apologies for freaking out.

File: hyperhtml.js (hunk 8 above)
Comment (WebReflection) [inline hyperhtml.js]: update: not a single `-1 < ...` left on my local code :wink:

File: hyperhtml.js (hunk 2 above)
Comment (WebReflection) [inline hyperhtml.js]: update: all `-1 < ...` now are `... !== -1`

File: hyperhtml.js (hunk 14 above)
Comment (marcoscaceres) [inline hyperhtml.js]: ok, makes sense.

File: hyperhtml.js (hunk 1 above)
Comment (marcoscaceres) [inline hyperhtml.js]: no, virtual seems ok given the explanation. Thanks for describing it (definitely keep the above to put into the docs).

File: hyperhtml.js (hunk 9 above)
Comment (marcoscaceres) [inline hyperhtml.js]: sounds good.

File: hyperhtml.js (hunk 6 above)
Comment (marcoscaceres) [inline hyperhtml.js]: Ok, I see.

File: hyperhtml.js (hunk 15 above)
Comment (marcoscaceres) [inline hyperhtml.js]: LOL, I had no idea about td just being ignored! That's insane.

Comment (marcoscaceres) [inline hyperhtml.js]: I guess the comment is, "Overcome limitation in HTML, whereby `HTMLTable*Element` that are not wrapped in a `HTMLTableElement` are ignored"

File: hyperhtml.js [hunk 19]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+  // internal signal to switch adoption
+  var notAdopting = true;
+
+  // ---------------------------------------------
+  // Template related utilities
+  // ---------------------------------------------
+
+  // given a unique template object
+  // create, parse, and store retrieved info
+  function createTemplate(template) {
```
Comment (marcoscaceres) [inline hyperhtml.js]: STOPPING HERE ON DAY 2. ... close to finishing, but starting to fade out again.

File: hyperhtml.js [hunk 20]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+    ) {
+      attribute = attributes[i];
+      if (attribute.value === value) {
+        paths.push(
+          Path(
+            'attr',
+            // with IE the order doesn't really matter
+            // as long as the right attribute is addressed
+            IE ?
+              node[GET_ATTRIBUTE_NODE](IEAttributes.shift()) :
```
Comment (WebReflection) [inline hyperhtml.js]: IEAttributes is populated while replacing comments in the RegExp.

The discovery order will NOT match the attributes order here, which is why it has to be drained to get the right attribute left to right.

I give you an example. `<div attr="${1}" whatever="0" id="${2}">` in IE will result into an attributes list of ['id', 'attr'] and it's unpredictable with `class` and any other possible special attribute per node.

The RegExp will replace, left to right, the `<div attr=<!--UID-->" whatever="0" id="<!--UID-->">` as tring, because IE also crashes if it has attributes as comment (Edge inherited the shuffled order issue but at least not the throwing error one so it needs this hack too).

At that point it doesn't matter which attribute, in which order, the UID content is discovered, what we want to map in there is the attribute node `attr` first and the `id` after.

TL;DR it's a hack that resolve IE problems and it's granted to work thanks to tests.

Yes, it is scary, because scary has always been IE in general.

File: hyperhtml.js [hunk 21]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+  // Node.CONSTANTS
+  // without assuming Node is globally available
+  // since this project is used on the backend too
+  var ELEMENT_NODE = 1;
+  var ATTRIBUTE_NODE = 2;
+  var TEXT_NODE = 3;
+  var COMMENT_NODE = 8;
+  var DOCUMENT_FRAGMENT_NODE = 11;
+
+  var GET_ATTRIBUTE_NODE = 'getAttributeNode';
```
Comment (WebReflection) [inline hyperhtml.js]: plays well with minifiers and it's used as constant in switch cases too. Now locally I also have `OWN_SVG_OWNER` with similar role.

File: hyperhtml.js [hunk 22]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+  // Firefox < 55 has non standard template literals.
+  // https://bugzilla.mozilla.org/show_bug.cgi?id=1108941
+  // TODO: is there any better way to feature detect this ?
+  var FF = typeof navigator === 'object' &&
+            /Firefox\/(\d+)/.test(navigator.userAgent) &&
+            parseFloat(RegExp.$1) < 55;
+
+  // If attributes order is shuffled, threat the browser differently
+  // Usually this is a well known IE only limitation but some older FF does the same.
+  var IE =  (function (p) {
```
Comment (WebReflection) [inline hyperhtml.js]: a `var` less ? :smile: ... OK

File: hyperhtml.js [hunk 23]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+          // no need to check upfront for the indexOf
+          set: function (obj, value) {
+            v[k.push(obj) - 1] = value;
+          }
+        };
+      } :
+      Map;
+
+  // TODO: which browser needs these partial polyfills here?
+  var isArray = Array.isArray || function (a) { return a instanceof Array; };
```
Comment (WebReflection) [inline hyperhtml.js]: The only Array I check are internals. There is no way an Array from another Realm can reach this code and `instanceof` is 1000X faster than brand check.

Brand check is necessary when you don't own the input. Here I do

Comment (WebReflection) [inline hyperhtml.js]: On a second though, if somebody passes different arrays that are not wires, then I won't control the input. OK, Changing ... (sadly)

File: hyperhtml.js [hunk 24]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+  var trim = EXPANDO.trim || function () { return this.replace(/^\s+|\s+$/g, ''); };
```
Comment (WebReflection) [inline hyperhtml.js]: EXPANDO can only be a string and it will always be a string because it's used as ... `expendo` property.

I am trying to keep an eye on the code size too :wink:

File: hyperhtml.js [hunk 25]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+    templateObjects = Object.create(null);
+    unique = function (template) {
+      var key = template.join(UID);
+      return templateObjects[key] ||
+            (templateObjects[key] = template);
+    };
+  }
+
+  // use native .append(...childNodes) where available
+  var appendNodes = 'append' in document ?
```
Comment (WebReflection) [inline hyperhtml.js]: You still seem to ignore the fact I don't give browser environment for granted

File: hyperhtml.js [hunk 26]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+          i = 0,
+          length = childNodes.length;
+          i < length; i++
+        ) {
+          node.appendChild(childNodes[i]);
+        }
+      };
+
+  // returns children or retrieve them in IE/Edge
+  var getChildren = 'children' in document ?
```
Comment (WebReflection) [inline hyperhtml.js]: no, specially because that would return `true` being that inherited.

The `in` check is fine.

Comment (WebReflection) [inline hyperhtml.js]: it's a one off, it doesn't compromise any performance at all whatsoever.

File: hyperhtml.js [hunk 27]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+      function (node) { return node.children; } :
```
Comment (WebReflection) [inline hyperhtml.js]: I don't understand this comment.

Here the catch: in IE the document and the document fragment does not expose `children` and there are other cases too.

Converting children to Array when useless is not what this refactoring is about.

I am trying to avoid every possible redundant operation.

`slice.call` is used only when necessary and converting all children that are **never mutated** from any fragment seems unnecessary overhead.

100% test coverage is there to avoid problems with that in the future :wink:

File: hyperhtml.js [hunk 28]
```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
...
+        }
+        return parentNode;
+      };
+
+  // fixes IE problems with comments
+  if (IE) {
+    var IEAttributes;
+    var no = new RegExp('([^\\S][a-z]+[a-z0-9_-]*=)([\'"])' + UIDC + '\\2', 'g');
+    var comments = function ($0, $1, $2) {
+          IEAttributes.push($1.slice(1, -1));
```
Comment (WebReflection) [inline hyperhtml.js]: uhm ... where?

File: hyperhtml.js (hunk 19 above)
Comment (WebReflection) [inline hyperhtml.js]: Hey @marcoscaceres I've actually rush-replied 'cause I was in the middle of other changes so please don't get my comments wrong, I'm just trying to explain.

Meanwhile I've tried to follow all your hints and put comments everywhere to avoid confusion or be scared so thanks a lot for the effort!

File: hyperhtml.js (hunk 4 above)
Comment (WebReflection) [inline hyperhtml.js]: if legacy, these are also hard to update so no legacy project would adopt hyperHTML ? :smiley:

File: hyperhtml.js (hunk 24 above)
Comment (WebReflection) [inline hyperhtml.js]: EXPANDO after minifying is 1 char, `""` is two :P ... but yes, eventually I'd do that instead of the `String.prototype.` circus :wink:

File: hyperhtml.js (hunk 8 above)
Comment (WebReflection) [inline hyperhtml.js]: actually I've just realized I've lost many changes ... something went wrong with the merge
